### PR TITLE
Producer CPP Canary

### PR DESCRIPTION
**Pull request by stefankiesz**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: canary/producer-cpp/src/CanaryLogs.cpp
Comment (disa6302) [inline canary/producer-cpp/src/CanaryLogs.cpp:55]: Replace all these with `LOG_x`. DLOGs wont get pushed to cloudwatch

File: canary/producer-cpp/.vscode/settings.json
Comment (disa6302) [inline canary/producer-cpp/.vscode/settings.json]: Remove file

File: canary/producer-cpp/CMake/Utilities.cmake
Comment (disa6302) [inline canary/producer-cpp/CMake/Utilities.cmake]: Remove file

File: canary/producer-cpp/CMakeLists.txt
Comment (disa6302) [inline canary/producer-cpp/CMakeLists.txt:2]: Remove if not required

File: canary/producer-cpp/CMakeLists.txt
Comment (disa6302) [inline canary/producer-cpp/CMakeLists.txt:11]: Remove if not required. Goes for all the commented out pieces.

File: canary/producer-cpp/CMakeLists.txt
Comment (disa6302) [inline canary/producer-cpp/CMakeLists.txt:18]: Fix the message. "Tell SDK to..."

File: canary/producer-cpp/README.md
Comment (disa6302) [inline canary/producer-cpp/README.md:1]: Just adding a note: Needs to be completed.

File: canary/producer-cpp/jobs/canary_seed.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/canary_seed.groovy:39]: producer cpp canary. Be specific. Same goes for runner job descriptions.

File: canary/producer-cpp/jobs/orchestrator.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/orchestrator.groovy:6]: Remember to change this to master before merging because the branch automatically gets deleted once the PR is merged.

File: canary/producer-cpp/jobs/orchestrator.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/orchestrator.groovy:15]: Is this needed for the canary?

File: canary/producer-cpp/jobs/orchestrator.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/orchestrator.groovy:126]: Remove commented out portions if not applicable please.

File: canary/producer-cpp/kvs_log_configuration
Comment (disa6302) [inline canary/producer-cpp/kvs_log_configuration:1]: This file is not necessary since we end up copying as part of CMakeLists.txt file. Remove the file.

**Review by disa6302 — COMMENTED** [review]
General feedback: Let us get rid of all the commented out portions of code in the files.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:171]: General guidance: Let us following one style of naming: either camel case: `persistedAckLatencyDatum` or `persisted_ack_latency_datum`. Let us mix it. Since most of the code is using former, please modify relevant function names and variable names to be all camel cases type.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:184]: Let us change to an async call. Use PutMetricDataAsync instead. We do not want to block on this. You can refer to producer C canary for reference.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:218]: This else portion does not make sense. I think what you are going with if it neither received or persisted ack. However, we still want to track other acks (error and buffered( even if we are not going to publish metrics for them. It can be a simple log line, but we need it. So when you modify this to switch case, let us add error ack and buffered ack type as well.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:248]: Why this? We dont need random data because we have our actual gstreamer source frame data, right? You need to append the metadata to the actual frame data.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:265]: Hmm. Doesnt make sense. Why do you need an address for lastKeyFrameTime? Why are you using address instead of pointer? It needs to be `PUINT64 lastKeyFrameTime`. And in the place where this is being called, you need to send address.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:291]: Can we modularize this function further to get stream metrics and client metrics respectively? It will be cleaner.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:330]: Number of errors need not be double. It can be UINT64. Only rate needs to be DOUBLE.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:357]: Replace with Async call.

Comment (disa6302) [reply canary/producer-cpp/src/Canary.cpp:357]: Also, what metric is being pushed here? Shouldnt this be done for every stream metric and other metrics?

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:591]: Use `GETENV` instead of `getenv`

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:573]: Think if can push this credential configuration within CanaryConfig.cpp. The purpose CanaryConfig.cpp should be to do this kind of env based config set up. Both static and IoT credential provider. See if it can be done as part of `initConfigWithEnvVars`

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:673]: Not needed at the moment. We do not support rtsp yet in our canaries.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:715]: Do we not need this? If not, remove.

File: canary/producer-cpp/src/Include.h
Comment (disa6302) [inline canary/producer-cpp/src/Include.h:62]: Remove if not required. Also, If we plan to have an Include.h, I would suggest moving #defines from all the files into this and have all header files reference Include.h file. It is more modular and cleaner.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:769]: Would suggest changing the place for definition of TEST_SOURCE outside CustomData to avoid this.

File: canary/producer-cpp/src/CanaryConfig.cpp
Comment (disa6302) [inline canary/producer-cpp/src/CanaryConfig.cpp:23]: Use PIC functions. GETENV instead of getenv.

File: canary/producer-cpp/src/CanaryConfig.cpp
Comment (disa6302) [inline canary/producer-cpp/src/CanaryConfig.cpp:31]: For string to int or other data type conversions, reference this: https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/utils/include/com/amazonaws/kinesis/video/utils/Include.h. There is a strtoint() function that does this.

File: canary/producer-cpp/src/CanaryLogs.cpp
Comment (disa6302) [inline canary/producer-cpp/src/CanaryLogs.cpp:93]: Remove if not needed?

File: canary/producer-cpp/src/Canary.cpp
Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp:265]: Isn't this a reference? https://www.ibm.com/docs/en/zos/2.4.0?topic=calls-pass-by-reference-c-only

File: canary/producer-cpp/src/Canary.cpp
Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp:347]: All the metrics are added to the cwRequest to be sent all together in the cwRequest, and it is done in four places in the code: upon persisted ack, upon received ack, upon keyframe, and for startup latency.

Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp:347]: (In the case above, it is for all the metrics upon keyframe)

Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp:347]: And for AmazonCloudWatchAsync calls I am getting: 
`error: no matching function for call to ‘Aws::CloudWatch::CloudWatchClient::PutMetricDataAsync(Aws::CloudWatch::Model::PutMetricDataRequest&)`
Is there a different namespace I should be using that includes this? Everything I'm seeing online on this is for Java.

File: canary/producer-cpp/src/Canary.cpp
Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp]: I moved all of the GETENV calls to CanaryConfig.cpp which cleaned it up some, but the rest of the function must stay outside of the CanaryConfig class because it involves CustomData members, and CanaryConfig does not have ability to access CustomData (only vis versa), I can make them point to each other if that would be better.

File: canary/producer-cpp/src/Canary.cpp
Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp]: Separated client from stream, and I also modularized error metrics push into its own function too.

File: canary/producer-cpp/src/Include.h
Comment (stefankiesz) [reply canary/producer-cpp/src/Include.h]: Having Include.h file in all the other headers would not work because they are all being aggregated to be included into the Include.h file - this aggregation is necessary because there are classes that refer to other classes.

File: canary/producer-cpp/src/Canary.cpp
Comment (stefankiesz) [reply canary/producer-cpp/src/Canary.cpp]: Fixed this for everything but the iot credential and canary callback stuff - seems to make more sense to leave as is for these two.

File: canary/producer-cpp/jobs/orchestrator.groovy
Comment (stefankiesz) [reply canary/producer-cpp/jobs/orchestrator.groovy:15]: Not needed, but it seemed like a useful thing to keep from the C Canary, it makes it easy to set the typical durations, delays, etc. I set the env vars for Canary duration this using these definitions

File: canary/producer-cpp/jobs/orchestrator.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/orchestrator.groovy:118]: As epr our discussion, can we reduce the number of builds to unique runs? As per this, we would only need a Realtime long run and realtime periodic. Also change the way the stream names are configured. Remove the `-01`, `-02`, etc.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:231]: Doing this would overwrite the gstreamer pFrame->frameData. You would have to memcopy the GStreamer frame data to the location starting from CANARY_METADATA_SIZE to ensure the header is appended at the start.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:278]: Remove.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:674]: Remove / uncomment.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:305]: Again. Not sure I understand this. Using pushMetric, you are setting cwRequest.AddMetricDatum = metricDatum. This call to `pushMetric` happens for every unique metric, however, your PutMetricDataAsync call is made once? Wouldnt the cwRequest get overwritten? With this call, it seems like you are actually only publishing only aggregated metrics if useAggMetrics is set and totalNumberOfErrors if not. You will need this line after every metric.

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:205]: Same about putMetricDataAsync. Please fix this everywhere. We need per stream metrics as well and not just aggregated. Include the call as part of `pushMetric` API so that the `PutMetricDataAsync` is invoked only at one location. Makes code cleaner.

File: canary/producer-cpp/jobs/canary_seed.groovy
Comment (disa6302) [inline canary/producer-cpp/jobs/canary_seed.groovy:11]: Namespace should be changed to producercpp

File: canary/producer-cpp/src/CanaryLogs.cpp
Comment (disa6302) [inline canary/producer-cpp/src/CanaryLogs.cpp:86]: Change from DLOGx to LOG_* .

File: canary/producer-cpp/jobs/canary_seed.groovy
Comment (stefankiesz) [reply canary/producer-cpp/jobs/canary_seed.groovy:11]: Just temporarily changed it back to producer for testing something

File: canary/producer-cpp/src/Canary.cpp
Comment (disa6302) [inline canary/producer-cpp/src/Canary.cpp:232]: You can avoid this by just memcpy'ing in `create_kinesis_video_frame`. frameData is being reinterpreted there. All you need to do is `MEMCPY(frame->frameData + METADATA_SIZE, frame->frameData, len) and set `frame->size` to include `METADATA_SIZE` and it will be handled. Then invoking this function is as it is done here. Your implementation would lead to memory leak because you do not clear the memory.

**Review by disa6302 — CHANGES_REQUESTED** [review]
Good job. Let us just fix the frame data populating portion. Also, please remove usage of couts and use the logger instead as added in previous comments :)

**Review by disa6302 — APPROVED** [review]
Yay!

**Pull request merged**
